posthog/async_migrations/migrations/0004_replicated_schema.py

```diff
@@ -219,35 +236,46 @@ def move_partitions(self, from_table: str, to_table: str):
             running_merges == 0
         ), f"No merges should be running on tables while partitions are being moved. table={from_table}"
 
-        partitions = sync_execute(
-            "SELECT DISTINCT partition FROM system.parts WHERE database = %(database)s AND table = %(table)s",
-            {"database": settings.CLICKHOUSE_DATABASE, "table": from_table},
-        )
-
-        for (partition,) in partitions:
-            logger.info("Moving partitions between tables", from_table=from_table, to_table=to_table, id=partition)
-            # :KLUDGE: Partition IDs are special and cannot be passed as arguments
-            sync_execute(f"ALTER TABLE {to_table} ATTACH PARTITION {partition} FROM {from_table}")
-            sync_execute(f"ALTER TABLE {from_table} DROP PARTITION {partition}")
+        retry = 0
+        while True:
+            partitions = sync_execute(
+                "SELECT DISTINCT partition FROM system.parts WHERE database = %(database)s AND table = %(table)s AND active",
+                {"database": settings.CLICKHOUSE_DATABASE, "table": from_table},
+            )
 
-    def rename_tables(
-        self, data_table_name, tmp_table_name, new_main_table_name, backup_table_name, skip_if_backup_exists=False
-    ):
+            try:
+                for (partition,) in partitions:
+                    logger.info(
+                        "Moving partitions between tables", from_table=from_table, to_table=to_table, id=partition
+                    )
+                    # :KLUDGE: Partition IDs are special and cannot be passed as arguments
+                    sync_execute(f"ALTER TABLE {to_table} REPLACE PARTITION {partition} FROM {from_table}")
+                    sync_execute(f"ALTER TABLE {from_table} DROP PARTITION {partition}")
+                break
+            # :KLUDGE: Logic to retry moving parts for person_distinct_id2 table
+            except ServerException as err:
+                error_code = lookup_error_code(err)
+                if error_code != "KEEPER_EXCEPTION" or retry > self.MOVE_PARTS_RETRIES:
```

Comment:
This means that we're going to do `self.MOVE_PARTS_RETRIES + 1` retries, which I find slightly confusing to read/reason about. Consider
```                
  retries_left = self.MOVE_PARTS_RETRIES
  while True:
    ...
    if error_code != "KEEPER_EXCEPTION" or not retries_left:  # or retries_left <= 0
       ...
    else:
      retries_left -= 1
```